components/places/src/bookmark_sync/store.rs

```diff
@@ -1138,6 +1139,69 @@ impl<'a> Merger<'a> {
         result
     }
 
+    /// Prepares synced bookmarks for merging.
+    fn prepare(&self) -> Result<()> {
+        // Like keywords, Sync associates tags with bookmarks, but Places
+        // associates them with URLs. This means multiple bookmarks with the
+        // same URL should have the same tags. In practice, different tags for
+        // bookmarks with the same URL are some of the most common validation
+        // errors we see.
+        //
+        // Unlike keywords, the relationship between URLs and tags in many-many:
+        // multiple URLs can have the same tag, and a URL can have multiple
+        // tags. So, to find mismatches, we need to compare the tags for each
+        // URL with the tags for each item.
+        //
+        // We could fetch both lists of tags, sort them, and then compare them.
+        // But there's a trick here: we're only interested in whether the tags
+        // _match_, not the tags themselves. So we sum the tag IDs!
```

Comment:
@eoger astutely asked if this would still work if one bookmark has tag IDs `[3, 4]`, and the second one has `[7]`. It would, and here's why: Let's say bookmarks A1 and A2 have the same URL A. `tagsByPlaceId` will have one row for the URL, `(A, 14)` (14 because 3 + 4 + 7—we're summing the tag IDs for the URL). `tagsByItemId` will have two rows, one for each item: `(A1, 7)` and `(A2, 7)`. So, even though the tag IDs sum up to 7 for both items, that sum doesn't match the sum for the URL A, and we'll flag A1 and A2 for reupload.

It's also why the `DISTINCT` is important for `tagsByPlaceId`. Let's say A1 and A2 have matching tags `[3, 4]`. In that case, `tagsByPlaceId` should have `(A, 7)`—we don't want to double-count the tags for the URL! And `tagsByItemId` will be `(A1, 7)` and `(A2, 7)`, so they'll match.

Reply:
Though, this made me think of a corner case: what if A1 has tags `[3, 4]` and A2 doesn't have _any_ tags? `tagsByPlaceId` won't catch it (thanks to the `HAVING COUNT(*) > 1`), and the inner `JOIN` to `tagsByItemId` means we'll filter out A2, oops!

So I tweaked the query to:

* Remove the `HAVING COUNT(*) > 1`. The join to `moz_bookmarks_synced_tag_relation` still means we'll filter out bookmarks without tags, and the query plan didn't change here.
* Use a `LEFT JOIN tagsByItemId t ON t.itemId = v.id`, to catch items without tags.
* Use `IS NOT` instead of `<>`, because `IS NOT` can compare nulls. (`NULL` in SQL is like `NaN` in JavaScript, it's toxic ☠️ `NULL` isn't equal or unequal to anything, including itself, so `1 <> NULL` is `NULL` (which is then cast to false 😱), not true...but `IS` handles it correctly: `1 IS NOT NULL` is true).

And added a test. Check out `bookmarkFFF1` and F2.